internal/services/compute/virtual_machine_implicit_data_disk_resource.go

```diff
@@ -0,0 +1,462 @@
+				string(virtualmachines.DiskCreateOptionTypesCopy),
+			}, false),
+		},
+
+		"delete_option": {
+			Type:     pluginsdk.TypeString,
+			Required: true,
+			ValidateFunc: validation.StringInSlice([]string{
+				string(virtualmachines.DiskDeleteOptionTypesDetach),
+				string(virtualmachines.DiskDeleteOptionTypesDelete),
+			}, false),
+		},
+
+		"disk_size_gb": {
+			Type:         pluginsdk.TypeInt,
+			Required:     true,
+			ForceNew:     true,
+			ValidateFunc: validation.IntBetween(1, 1023),
+		},
+
+		"source_resource_id": {
+			Type:         pluginsdk.TypeString,
+			Required:     true,
+			ForceNew:     true,
+			ValidateFunc: validation.Any(snapshots.ValidateSnapshotID, commonids.ValidateManagedDiskID),
+		},
+
+		"write_accelerator_enabled": {
+			Type:     pluginsdk.TypeBool,
+			Optional: true,
+			Default:  false,
+		},
+	}
+}
+
+func (r VirtualMachineImplicitDataDiskResource) Attributes() map[string]*pluginsdk.Schema {
+	return map[string]*pluginsdk.Schema{}
+}
+
+func (r VirtualMachineImplicitDataDiskResource) Create() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 30 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			client := metadata.Client.Compute.VirtualMachinesClient
+
+			var config VirtualMachineImplicitDataDiskResourceModel
+			if err := metadata.Decode(&config); err != nil {
+				return fmt.Errorf("decoding: %+v", err)
+			}
+
+			subscriptionId := metadata.Client.Account.SubscriptionId
+			virtualMachineId, err := virtualmachines.ParseVirtualMachineID(config.VirtualMachineId)
+			if err != nil {
+				return err
+			}
+
+			locks.ByName(virtualMachineId.VirtualMachineName, VirtualMachineResourceName)
+			defer locks.UnlockByName(virtualMachineId.VirtualMachineName, VirtualMachineResourceName)
+
+			id := parse.NewDataDiskID(subscriptionId, virtualMachineId.ResourceGroupName, virtualMachineId.VirtualMachineName, config.Name)
+
+			resp, err := client.Get(ctx, *virtualMachineId, virtualmachines.DefaultGetOperationOptions())
+			if err != nil {
+				return fmt.Errorf("checking for the presence of an existing %s: %+v", *virtualMachineId, err)
+			}
+
+			expandedDisk := virtualmachines.DataDisk{
+				Name:         pointer.To(config.Name),
+				Caching:      pointer.To(virtualmachines.CachingTypes(config.Caching)),
+				CreateOption: virtualmachines.DiskCreateOptionTypes(config.CreateOption),
+				DeleteOption: pointer.To(virtualmachines.DiskDeleteOptionTypes(config.DeleteOption)),
+				DiskSizeGB:   pointer.To(config.DiskSizeGb),
+				Lun:          config.Lun,
+				SourceResource: &virtualmachines.ApiEntityReference{
+					Id: pointer.To(config.SourceResourceId),
+				},
+				WriteAcceleratorEnabled: pointer.To(config.WriteAcceleratorEnabled),
+			}
+
+			if model := resp.Model; model != nil {
+				if props := model.Properties; props != nil {
+					if profile := props.StorageProfile; profile != nil {
+						if dataDisks := profile.DataDisks; dataDisks != nil {
```

Comment:
there's ways to provision a VM without a StorageProfile and/or DataDisks (e.g. when creating from certain types of image) - as such we'd end up doing nothing in those cases here - we'd need to update this to account for that

Reply:
code updated. Thanks.